### PR TITLE
ticdc(codec): add avro refactor design

### PR DESCRIPTION
**Pull request by zhangyangyu**
### What problem does this PR solve?

Issue Number: ref #5338 

### Release note 

```release-note
None
```

### CODE REVIEW

File: docs/design/2022-05-12-ticdc-avro-protocol-refactor.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+## Detailed Design
+
+### New Config Items
+
+| Config item                        | Option values          | Default | Explain                                                                                                                                                                                                                                                                                                         |
+| ---------------------------------- | ---------------------- | ------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| protocol                           | canal-json / flat-avro | -       | Specify the message format which output to the kafka.<br>The `flat-avro` option means that using the avro format design by this document.                                                                                                                                                                       |
```
Comment (zhaoxinyu) [inline docs/design/2022-05-12-ticdc-avro-protocol-refactor.md:47]:
```suggestion
| protocol                           | canal-json / flat-avro | -       | Specify the format of messages written to Kafka.<br>The `flat-avro` option means that using the avro format design by this document.                                                                                                                                                                       |
```

File: docs/design/2022-05-12-ticdc-avro-protocol-refactor.md [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+| schema-registry                    | -                      | -       | Specifies the schema registry endpoint.                                                                                                                                                                                                                                                                         |
+| avro-decimal-handling-mode         | precise / string       | precise | Specifies how the TiCDC should handle values for DECIMAL columns:<br>`precise` option represents encoding decimals as precise bytes.<br>`string` option encodes values as formatted strings, which is easy to consume but semantic information about the real type is lost.                                     |
+| avro-bigint-unsigned-handling-mode | long / string          | long    | Specifies how the TiCDC should handle values for UNSIGNED BIGINT columns:<br>`long` represents values by using Avro long(64-bit signed integer) which might overflow but which is easy to use in consumers.<br>`string` represents values by string which is precision but which is need to parse by consumers. |
```
Comment (zhaoxinyu) [inline docs/design/2022-05-12-ticdc-avro-protocol-refactor.md:51]:
```suggestion
| avro-bigint-unsigned-handling-mode | long / string          | long    | Specifies how the TiCDC should handle values for UNSIGNED BIGINT columns:<br>`long` represents values by using Avro long(64-bit signed integer) which might overflow but which is easy to use in consumers.<br>`string` represents values by string which is precise but which needs to be parsed by consumers. |
```

File: docs/design/2022-05-12-ticdc-avro-protocol-refactor.md [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+            "name":"_tidb_commit_physical_time",
+            "type":"long"
+        }
+    ]
+}
+```
+
+- `{{RecordName}}` represents full qualified table name.
+- `{{ColumnValueBlock}}` represents a JSON block, which defines a column value of a key.
+- `_tidb_op` used to distinguish between INSERT or UPDATE events, optional values are "c" / "u".
```
Comment (zhaoxinyu) [inline docs/design/2022-05-12-ticdc-avro-protocol-refactor.md:103]:
```suggestion
- `_tidb_op` is used to distinguish between INSERT or UPDATE events, optional values are "c" / "u".
```

File: docs/design/2022-05-12-ticdc-avro-protocol-refactor.md [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+- Insert row and check the row in downstream database.
+- Update row and check the row in downstream database.
+- Delete row and check the row in downstream database.
+
+#### Schema Tests
+
+- When the schema is not in schema registry, a fresh new schema is created.
+- When the schema is in schema registry and pass compatibility check, a new version is created.
+- When the schema is in schema registry and not pass compatibility check, reports error.
```
Comment (zhaoxinyu) [inline docs/design/2022-05-12-ticdc-avro-protocol-refactor.md:184]: ### 
```suggestion
- When the schema is in schema registry and cannot pass compatibility check, reports error.
```

**Review by leoppro — APPROVED** [review]
LGTM

**Pull request merged**